src/raw-value.ts

```diff
@@ -126,64 +78,136 @@ function createReverseShadowTarget(target: ReverseProxyTarget): ReverseShadowTar
+            secPartialDesc.get = env.getSecureValue(secPartialDesc.get);
+        }
+        return secPartialDesc;
+    }
+
+    function lockShadowTarget(shadowTarget: ReverseShadowTarget, originalTarget: ReverseProxyTarget) {
+        // todo: leaking symbols
+        const targetKeys = ownKeys(originalTarget);
+        for (let i = 0, len = targetKeys.length; i < len; i += 1) {
+            const key = targetKeys[i];
+            const rawDesc = ReflectGetOwnPropertyDescriptor(shadowTarget, key);
+            if (isUndefined(rawDesc) || rawDesc.configurable === true) {
+                const secDesc = ReflectGetOwnPropertyDescriptor(originalTarget, key) as PropertyDescriptor;
+                ReflectDefineProperty(shadowTarget, key, getRawDescriptor(secDesc));
             }
         }
+        ReflectPreventExtensions(shadowTarget);
     }
 
     class ReverseProxyHandler implements ProxyHandler<ReverseProxyTarget> {
         // original target for the proxy
         private readonly target: ReverseProxyTarget;
-        // metadata about the shape of the target
-        private readonly meta: TargetMeta;
 
-        constructor(target: ReverseProxyTarget, meta: TargetMeta) {
+        constructor(target: ReverseProxyTarget) {
             this.target = target;
-            this.meta = meta;
-        }
-        initialize(shadowTarget: ReverseShadowTarget) {
-            const { meta } = this;
-            // adjusting the proto chain of the shadowTarget (recursively)
-            const rawProto = env.getRawValue(meta.proto);
-            ReflectSetPrototypeOf(shadowTarget, rawProto);
-            // defining own descriptors
-            copyReverseOwnDescriptors(shadowTarget, meta.descriptors);
-            // reserve proxies are always frozen
-            freeze(shadowTarget);
             // future optimization: hoping that proxies with frozen handlers can be faster
             freeze(this);
         }
-        apply(shadowTarget: ReverseShadowTarget, thisArg: RawValue, argArray: RawValue[]): RawValue {
+        get(shadowTarget: ReverseShadowTarget, key: PropertyKey, receiver: RawObject): SecureValue {
+            const { target } = this;
+            // todo: leaking key as symbol
+            const desc = getPropertyDescriptor(target, key);
+            if (isUndefined(desc)) {
+                return desc;
+            }
+            const { get } = desc;
+            if (isFunction(get)) {
+                // calling the getter with the raw receiver because the getter expects a raw value
+                // it also returns a raw value
+                return apply(env.getRawValue(get), receiver, emptyArray);
+            }
+            return env.getRawValue(desc.value);
+        }
+        set(shadowTarget: ReverseShadowTarget, key: PropertyKey, value: RawValue, receiver: RawObject): boolean {
+            const { target } = this;
+            // todo: leaking key as symbol
+            const secValue = env.getSecureValue(value);
```

Comment:
👆 is there a reason to hoist the `secValue` assignment when it's used at the bottom of the function. I'm wondering if we can avoid the added lookup for the cases that exit early (`return` early).

Reply:
will fix it... probably result of some refactors